File: iosApp/Views/Settings/ModelsSection.swift
import SwiftUI
@preconcurrency import Shared

/// Models section (spec §11). Three groups:
///
/// 1. Bundled — every `bundled = true` entry from `model_registry.json`.
///    Marked "Available offline" with a check.
/// 2. Downloaded — models currently cached in the application support directory.
/// 3. Available — every other non-bundled entry, downloadable from HF.
///
/// A "Clear all caches" action is enabled when at least one model is downloaded.
struct ModelsSection: View {

    @Environment(ModelRegistryService.self) private var modelRegistry
    @Environment(ModelDownloadService.self) private var modelDownloadService
    @Environment(SettingsStore.self) private var settings

    @State private var showClearConfirmation = false

    var body: some View {
        Group {
            // MARK: Bundled
            Section {
                ForEach(bundledEntries, id: \.id) { entry in
                    HStack {
                        VStack(alignment: .leading) {
                            Text(entry.displayName)
                            Text(entry.id)
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                        Spacer()
                        Label("Available offline", systemImage: "checkmark.circle.fill")
                            .labelStyle(.iconOnly)
                            .foregroundStyle(.green)
                    }
                }
            } header: {
                Text("Bundled")
            } footer: {
                Text("Bundled models ship inside the app and run without network access.")
            }

            // MARK: Downloaded
            Section {
                if downloadedEntries.isEmpty {
                    Text("No downloaded models")
                        .foregroundStyle(.secondary)
                        .italic()
                } else {
                    ForEach(downloadedEntries, id: \.id) { entry in
                        downloadedRow(entry: entry)
                    }
                }
            } header: {
                HStack {
                    Text("Downloaded")
                    if modelDownloadService.downloadedModelCount > 0 {
                        Text("\(modelDownloadService.downloadedModelCount)")
                            .font(.caption2)
                            .padding(.horizontal, 5)
                            .padding(.vertical, 1)
                            .background(Color.blue.opacity(0.2))
                            .foregroundStyle(.blue)
                            .clipShape(Capsule())
                    }
                }
            } footer: {
                Text("Downloaded models are stored in the app sandbox and survive app restarts.")
            }

            // MARK: Available
            Section {
                ForEach(availableEntries, id: \.id) { entry in
                    availableRow(entry: entry)
                }
            } header: {
                Text("Available")
            } footer: {
                cacheFooter
            }
        }
        .confirmationDialog(
            "Clear all downloaded models?",
            isPresented: $showClearConfirmation,
            titleVisibility: .visible
        ) {
            Button("Clear all caches", role: .destructive) {
                modelDownloadService.clearAllCaches(settings: settings)
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("This will delete all downloaded models from the app sandbox. The bundled BNLeaky_Keras model will remain available.")
        }
    }

    // MARK: - Computed entry lists

    private var bundledEntries: [ModelRegistryEntry] {
        modelRegistry.registry.all().filter { $0.bundled }
    }

    /// Non-bundled entries that are in .downloaded state.
    private var downloadedEntries: [ModelRegistryEntry] {
        modelRegistry.registry.all().filter { entry in
            !entry.bundled &&
            modelDownloadService.downloadStates[entry.id] == .downloaded
        }
    }

    /// Non-bundled entries that are NOT in .downloaded state (available for download).
    private var availableEntries: [ModelRegistryEntry] {
        modelRegistry.registry.all().filter { entry in
            guard !entry.bundled else { return false }
            switch modelDownloadService.downloadStates[entry.id] ?? .notDownloaded {
            case .downloaded: return false
            default: return true
            }
        }
    }

    // MARK: - Row builders

    @ViewBuilder
    private func downloadedRow(entry: ModelRegistryEntry) -> some View {
        HStack {
            VStack(alignment: .leading) {
                Text(entry.displayName)
                Text(entry.id)
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
            Spacer()
            Label("Downloaded", systemImage: "arrow.down.circle.fill")
                .labelStyle(.iconOnly)
                .foregroundStyle(.blue)
            Button {
                modelDownloadService.deleteModel(modelId: entry.id)
            } label: {
                Image(systemName: "trash")
                    .foregroundStyle(.red)
            }
            .buttonStyle(.plain)
        }
    }

    @ViewBuilder
    private func availableRow(entry: ModelRegistryEntry) -> some View {
        let state = modelDownloadService.downloadStates[entry.id] ?? .notDownloaded
        switch state {
        case .notDownloaded:
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text(entry.displayName)
                    Text(String(format: "%.1f MB • Requires internet", entry.iosFileSizeMb))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
                Spacer()
                Button("Download") {
                    Task { await modelDownloadService.download(entry: entry) }
                }
                .buttonStyle(.bordered)
                .controlSize(.small)
            }

        case .downloading(let progress):
            VStack(alignment: .leading, spacing: 6) {
                Text(entry.displayName)
                ProgressView(value: progress)
                let receivedMb = entry.iosFileSizeMb * progress
                Text(String(format: "%.1f MB of %.1f MB", receivedMb, entry.iosFileSizeMb))
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

        case .compiling:
            VStack(alignment: .leading, spacing: 6) {
                Text(entry.displayName)
                ProgressView()
                Text("Compiling…")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

        case .downloaded:
            // Should not appear here (covered by downloadedEntries), but handle defensively.
            EmptyView()

        case .failed(let message):
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text(entry.displayName)
                    Text(message)
                        .font(.caption)
                        .foregroundStyle(.red)
                }
                Spacer()
                Button("Retry") {
                    Task { await modelDownloadService.download(entry: entry) }
                }
                .buttonStyle(.bordered)
                .controlSize(.small)
                .tint(.orange)
            }
        }
    }

    // MARK: - Footer

    private var cacheSizeMb: Double {
        downloadedEntries.reduce(0) { $0 + $1.iosFileSizeMb }
    }

    private var cacheFooter: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(String(format: "Total cache size: %.1f MB", cacheSizeMb))
                .font(.caption)
            Button(role: .destructive) {
                showClearConfirmation = true
            } label: {
                Text("Clear all caches")
            }
            .disabled(!modelDownloadService.hasDownloadedModels)
        }
    }
}

File: androidApp/src/main/kotlin/com/malaria/android/ui/settings/ModelsSection.kt
package com.malaria.android.ui.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.malaria.android.services.ModelDownloadState
import com.malaria.android.ui.locals.LocalModelDownloadService
import com.malaria.android.ui.locals.LocalModelRegistry
import com.malaria.android.ui.locals.LocalSettingsStore
import com.malaria.registry.ModelRegistryEntry

/**
 * Models section (spec §11). Mirrors `ModelsSection.swift`. Shows bundled,
 * downloaded, and available models with real download counts and a
 * working "Clear all caches" button.
 */
@Composable
fun ModelsSection() {
    val modelRegistry = LocalModelRegistry.current
    val modelDownloadService = LocalModelDownloadService.current
    val settings = LocalSettingsStore.current
    val downloadStates by modelDownloadService.downloadStates.collectAsStateWithLifecycle()

    var entries by remember { mutableStateOf<List<ModelRegistryEntry>>(emptyList()) }
    var showClearDialog by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        entries = runCatching { modelRegistry.registry().all() }.getOrDefault(emptyList())
    }

    val bundled = entries.filter { it.bundled }
    val available = entries.filter { !it.bundled }
    val downloaded = available.filter { downloadStates[it.id] is ModelDownloadState.Downloaded }
    val cacheSizeMb = modelDownloadService.totalCacheSizeMb
    val hasCache = downloaded.isNotEmpty()

    Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
        SectionScaffold(
            header = "Models — Bundled",
            footer = "Bundled models ship inside the app and run without network access.",
        ) {
            bundled.forEach { entry ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Column(modifier = Modifier.padding(end = 12.dp)) {
                        Text(entry.displayName, style = MaterialTheme.typography.bodyMedium)
                        Text(
                            text = entry.id,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Icon(
                        imageVector = Icons.Default.CheckCircle,
                        contentDescription = "Available offline",
                        tint = Color(0xFF2E7D32),
                    )
                }
            }
        }

        SectionScaffold(
            header = "Models — Downloaded",
            footer = "Downloaded models are stored in the app sandbox and survive app restarts.",
        ) {
            if (downloaded.isEmpty()) {
                Text(
                    text = "No downloaded models",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            } else {
                downloaded.forEach { entry ->
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceBetween,
                    ) {
                        Column {
                            Text(entry.displayName, style = MaterialTheme.typography.bodyMedium)
                            Text(
                                text = String.format("%.1f MB", entry.androidFileSizeMb),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        Icon(
                            imageVector = Icons.Default.CheckCircle,
                            contentDescription = "Downloaded",
                            tint = MaterialTheme.colorScheme.primary,
                        )
                    }
                }
            }
        }

        SectionScaffold(header = "Models — Available") {
            available.forEach { entry ->
                Column(modifier = Modifier.fillMaxWidth()) {
                    Text(entry.displayName, style = MaterialTheme.typography.bodyMedium)
                    Text(
                        text = "Requires internet · ${String.format("%.1f MB", entry.androidFileSizeMb)}",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                Text(
                    text = if (hasCache) "Cache: ${String.format("%.1f MB", cacheSizeMb)}"
                    else "Total cache size: 0 MB",
                    style = MaterialTheme.typography.bodySmall,
                )
                OutlinedButton(
                    onClick = { showClearDialog = true },
                    enabled = hasCache,
                    modifier = Modifier.fillMaxWidth(),
                ) {
                    Text("Clear all caches")
                }
                if (!hasCache) {
                    Text(
                        text = "No cached models.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }
    }

    if (showClearDialog) {
        AlertDialog(
            onDismissRequest = { showClearDialog = false },
            title = { Text("Clear all caches?") },
            text = { Text("All downloaded models will be deleted. The default model will revert to BN + LeakyReLU (Keras).") },
            confirmButton = {
                TextButton(onClick = {
                    modelDownloadService.clearAllCaches(settings)
                    showClearDialog = false
                }) { Text("Clear") }
            },
            dismissButton = {
                TextButton(onClick = { showClearDialog = false }) { Text("Cancel") }
            },
        )
    }
}
